internal/pkg/service/appproxy/http/router.go

```diff
@@ -92,22 +96,43 @@ func (r *Router) CreateHandler() http.Handler {
 			return
 		}
 
-		// Delete all X-Kbc-* headers as they're used for internal information.
+		// Delete all X-KBC-* headers as they're used for internal information.
 		for name := range req.Header {
-			if strings.HasPrefix(name, "X-Kbc-") {
+			if strings.HasPrefix(strings.ToLower(name), "x-kbc-") {
 				req.Header.Del(name)
 			}
 		}
 
-		appID := AppID(appIDString.Emit())
+		appID := appIDString.Emit()
 
-		if handler, found := r.handlers[appID]; found {
-			handler.ServeHTTP(w, req)
-		} else {
-			r.logger.Infof(req.Context(), `application "%s" not found`, appID)
-			w.WriteHeader(http.StatusNotFound)
-			fmt.Fprintf(w, `Application "%s" not found.`, appID)
+		// Load configuration for given app.
+		config, err := r.loader.LoadConfig(req.Context(), appID)
+		if err != nil {
+			var sandboxesError *appconfig.SandboxesError
+			errors.As(err, &sandboxesError)
+			if sandboxesError != nil && sandboxesError.StatusCode() == http.StatusNotFound {
+				r.logger.Infof(req.Context(), `application "%s" not found`, appID)
+				w.WriteHeader(http.StatusNotFound)
+				fmt.Fprintf(w, `Application "%s" not found.`, appID)
+				return
+			}
+
+			w.WriteHeader(http.StatusServiceUnavailable)
+			fmt.Fprintf(w, `Unable to loading configuration for app "%s".`, appID)
+			return
+		}
+
+		// Recreate app handler if configuration changed.
+		handler, ok := r.appHandlers[appID]
+		if !ok || handler.Etag != config.ETag {
+			handler = appHandler{
+				Handler: r.createDataAppHandler(req.Context(), config),
+				Etag:    config.ETag,
+			}
+			r.appHandlers[appID] = handler
```

Comment:
Isn't there a lock needed? (`sync.Mutex/RWMutex`)

Reply:
`sync.Map` is not generic, so it requires re-typing, therefore it is used very rarely.

Reply:
Because you don't want to see these errors in production. :slightly_smiling_face:

```
fatal error: concurrent map read and map
```

https://stackoverflow.com/questions/45585589/golang-fatal-error-concurrent-map-read-and-map-write

Reply:
> Oh I thought you wanted the lock around

That would be nice too, but it will work without it. But the lock I'm writing about is essential.

Reply:
Because you probably don't have a test that sends parallel request and change the configuration of the app at the same time. (?)

Reply:
- Sending 100/1000 parallel requests is easy (100 goroutines + waitgroup).
- It is more difficult to figure out the app configuration change: there may be some Sandboxes API mock that will always send a random ETag and thus re-configuration will always happen? (or every `count % 5 ==0` times), or something similar.